frame/support/procedural/src/construct_runtime/parse.rs

```diff
@@ -143,81 +180,128 @@ pub struct PalletDeclaration {
 	pub index: Option<u8>,
 	pub path: PalletPath,
 	pub instance: Option<Ident>,
-	pub pallet_parts: Vec<PalletPart>,
+	pub pallet_parts: Option<Vec<PalletPart>>,
+	pub specified_parts: SpecifiedParts,
+}
+
+/// The possible declaration of pallet parts to use.
+#[derive(Debug, Clone)]
+pub enum SpecifiedParts {
+	/// Use all the pallet parts except those specified.
+	Exclude(Vec<PalletPartNoGeneric>),
+	/// Use only the specified pallet parts.
+	Use(Vec<PalletPartNoGeneric>),
+	/// use the all the pallet parts.
+	All,
 }
 
 impl Parse for PalletDeclaration {
 	fn parse(input: ParseStream) -> Result<Self> {
 		let name = input.parse()?;
 		let _: Token![:] = input.parse()?;
 		let path = input.parse()?;
-		let instance = if input.peek(Token![<]) {
+
+		// Parse for instance.
+		let instance = if input.peek(Token![::]) && input.peek3(Token![<]) {
+			let _: Token![::] = input.parse()?;
 			let _: Token![<] = input.parse()?;
 			let res = Some(input.parse()?);
 			let _: Token![>] = input.parse()?;
-			let _: Token![::] = input.parse()?;
 			res
+		} else if !(input.peek(Token![::]) && input.peek3(token::Brace)) &&
+			!input.peek(keyword::exclude_parts) &&
+			!input.peek(keyword::use_parts) &&
+			!input.peek(Token![=]) &&
+			!input.peek(Token![,]) &&
+			!input.is_empty()
+		{
+			return Err(input.error(
+				"Unexpected tokens, expected one of `::$ident` `::{`, `exclude_parts`, `use_parts`, `=`, `,`",
```

Comment:
Can we use `input.lookahead()` here to generate the error?

Reply:
the thing is that we look at 2 tokens and `lookahead2` doesn't exist

Reply:
Looking at one token at a time may actually be what we want though? Because the parser at this state accepts `::`, `exclude_parts`, `use_parts`, `=` or `,` as the next token, and if we want to check for what comes after a `::`, we would then explicitly check for `$ident` and `{` inside of a conditional branch that has already checked for the existence of `::`.

Reply:
I mean the error message will be more confusing if somebody writes: `frame_system:: exclude_parts {},` it will say that after the `::` it expects an ident or a `{`, but actually the user just doesn't want to write `::` in the first place